outlines/fsm/fsm.py

```diff
@@ -119,9 +140,63 @@ def __init__(self, regex_string: str, tokenizer: "Tokenizer"):
         self.final_states = regex_fsm.finals | {
             -1
         }  # Include the EOS token in final states
-        self.vocabulary = tokenizer.vocabulary.values()
+        self.tokenizer = tokenizer
+        self.vocabulary = tokenizer.vocabulary
         self.end_token_id = tokenizer.eos_token_id
 
+    def align_prompt_tokens(self, prompt: str) -> str:
+        """Remove the last token from the prompt and update the states_to_token_maps accordingly"""
+        token_ids, _ = self.tokenizer.encode(prompt)
+        last_token_id = int(token_ids[0][-1])
+        last_token_text = self.tokenizer.decode([last_token_id])[0]
+        last_token_length = len(last_token_text)
+        # select the tokens that start with the text removed from the prompt
+        crossing_tokens = {
+            token: text
+            for text, token in self.vocabulary.items()
+            if text.startswith(last_token_text)
+        }
+        # keep only the tokens whose text after the boundary matches the fsm
+        valid_tokens_states = self.find_valid_crossing_tokens(
+            crossing_tokens, last_token_length
+        )
+        # update the states_to_token_maps in the following manner:
+        # the value of the starting state is assigned to a new state, the starting state is now the
+        # valid_tokens_states found above
+        additional_state_id = (
+            max(list(self.states_to_token_maps.keys()) + list(self.final_states)) + 1
+        )
+        self.states_to_token_maps[additional_state_id] = self.states_to_token_maps[0]
+        self.states_to_token_maps[0] = {}
+        for token, state in valid_tokens_states:
+            if state == 0:
+                self.states_to_token_maps[0][token] = additional_state_id
+            else:
+                self.states_to_token_maps[0][token] = state
+        return prompt[: -len(last_token_text)]
+
+    def find_valid_crossing_tokens(
+        self, crossing_tokens: Dict[int, str], last_token_length: int
+    ) -> List[Tuple[int, int]]:
+        """For each crossing token, check that the characters after the boundary match the FSM
+        and find the state it would lead to. Return the valid tokens with the associated state
+        """
+        valid_tokens = []
+        for token, text in crossing_tokens.items():
+            is_valid = True
+            crossing_text = text[last_token_length:]
+            state = 0
+            for char in crossing_text:
+                char_token = self.vocabulary.get(char)
+                try:
+                    state = self.states_to_token_maps[state][char_token]  # type: ignore
+                except KeyError:
+                    is_valid = False
+                    break
```

Comment:
Yes, this works because of the way tokenizers are trained, if a given token is present at any given transition then the individual characters that make up this token are also necessarily present.

We should add a word of warning here.